File: ios/Classes/TimelineEditModel.swift
import Foundation

struct TimelineEditSnapshot {
  let clips: [TimelineClipDescriptor]
  let audioTrack: AudioTrackDescriptor?
}

final class TimelineEditModel {
  private let limit: Int
  private var undoStack: [TimelineEditSnapshot] = []
  private var redoStack: [TimelineEditSnapshot] = []

  init(limit: Int = 50) {
    self.limit = max(limit, 1)
  }

  var canUndo: Bool {
    return !undoStack.isEmpty
  }

  var canRedo: Bool {
    return !redoStack.isEmpty
  }

  func pushSnapshot(_ snapshot: TimelineEditSnapshot) {
    undoStack.append(snapshot)
    if undoStack.count > limit {
      undoStack.removeFirst(undoStack.count - limit)
    }
    redoStack.removeAll()
  }

  func undo(current: TimelineEditSnapshot) -> TimelineEditSnapshot? {
    guard let previous = undoStack.popLast() else {
      return nil
    }
    redoStack.append(current)
    if redoStack.count > limit {
      redoStack.removeFirst(redoStack.count - limit)
    }
    return previous
  }

  func redo(current: TimelineEditSnapshot) -> TimelineEditSnapshot? {
    guard let next = redoStack.popLast() else {
      return nil
    }
    undoStack.append(current)
    if undoStack.count > limit {
      undoStack.removeFirst(undoStack.count - limit)
    }
    return next
  }
}

File: android/src/main/kotlin/com/andre/video_ultra_player/TimelineEditModel.kt
package com.andre.video_ultra_player

internal data class TimelineEditSnapshot(
    val clips: List<TimelineClip>,
    val audioTrack: AudioTrackDescriptor?,
    val textOverlays: List<TextOverlayDescriptor>
)

internal class TimelineEditModel(
    private val limit: Int = 50
) {
    private val undoStack = ArrayDeque<TimelineEditSnapshot>()
    private val redoStack = ArrayDeque<TimelineEditSnapshot>()
    private val safeLimit = limit.coerceAtLeast(1)

    val canUndo: Boolean
        get() = undoStack.isNotEmpty()

    val canRedo: Boolean
        get() = redoStack.isNotEmpty()

    fun pushSnapshot(snapshot: TimelineEditSnapshot) {
        undoStack.addLast(snapshot)
        while (undoStack.size > safeLimit) {
            undoStack.removeFirst()
        }
        redoStack.clear()
    }

    fun undo(current: TimelineEditSnapshot): TimelineEditSnapshot? {
        val previous = undoStack.removeLastOrNull() ?: return null
        redoStack.addLast(current)
        while (redoStack.size > safeLimit) {
            redoStack.removeFirst()
        }
        return previous
    }

    fun redo(current: TimelineEditSnapshot): TimelineEditSnapshot? {
        val next = redoStack.removeLastOrNull() ?: return null
        undoStack.addLast(current)
        while (undoStack.size > safeLimit) {
            undoStack.removeFirst()
        }
        return next
    }
}
